File: src/0002_add_two_numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carryover = 0
        anchor = ListNode()
        current = anchor
        
        while l1 or l2 or carryover:
            res = (l1.val if l1 else 0) + (l2.val if l2 else 0) + carryover
            carryover = res // 10
            value = res % 10
            current.next = ListNode(value)
            current = current.next 
            
            if l1: l1 = l1.next
            if l2: l2 = l2.next

        return anchor.next

File: src/0002_add_two_numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carryover = 0;
        ListNode* anchor = new ListNode();
        ListNode* tail = anchor;

        // while (l1 or l2 or (carryover != 0)) {
        //     int val1 = (l1) ? l1->val : 0;
        //     int val2 = (l2) ? l2->val : 0;
        while (l1 != nullptr || l2 != nullptr || carryover > 0) {
            int val1 = (l1 != nullptr) ? l1->val : 0;
            int val2 = (l2 != nullptr) ? l2->val : 0;
            int sum = val1 + val2 + carryover;
            int val = sum % 10;
            carryover = sum / 10;

            ListNode* nextNode = new ListNode(val);
            tail->next = nextNode;
            tail = tail->next;

            if (l1 != nullptr) l1 = l1->next;
            if (l2 != nullptr) l2 = l2->next;
        }

        ListNode* result = anchor->next;
        delete anchor;
        return result;
    }
};